### PR TITLE
Represent singularity-safe propagators for IAF models as classes

### PR DESCRIPTION
**Pull request by heplesser**
This PR is a follow-up replacing #2356. The content is the same except for the last few commits, which change the class names to `IAFPropagator{Exp,Alpha}`, make some changes to what is inlined and what not, and update documentation.

#2294 depends on this one.

### CODE REVIEW

File: libnestutil/iaf_propagator.cpp
Comment (clinssen) [inline libnestutil/iaf_propagator.cpp:111]: Could you make this 0.1 a ``const static double`` member of the class and add it to the class docstring?

**Review by clinssen — APPROVED** [review]
Approving aside from the one minor comment. Cheers!

**Comment by heplesser** [comment]
Turned into draft, realised some additional opportunities to tidy up code and documentation.

File: libnestutil/iaf_propagator.cpp
Comment (heplesser) [reply libnestutil/iaf_propagator.cpp]: I now moved this test into a method and added a comment pointing out that the precise value of the limit plays little role.

**Comment by heplesser** [comment]
@jougs @clinssen If think this PR is now ready for proper review. I am reasonably content with the class structure now. Please also take look at the Jupyter notebook. Once this is merged, #2294 can also finally be merged.

@jessica-mitchell I would like to provide links between [rst documentation (bottom of file)](https://github.com/heplesser/nest-simulator/blob/253986423a9bbf4f3ede8d70dc8e5d751af7ee20/doc/htmldoc/neurons/exact-integration.rst) and [Jupyter documentation (top of notebook)](https://github.com/heplesser/nest-simulator/blob/253986423a9bbf4f3ede8d70dc8e5d751af7ee20/doc/htmldoc/neurons/model_details/IAF_Integration_Singularity.ipynb). Do you have any suggestion about how to best do that?

**Comment by heplesser** [comment]
I will come with another update in a few minutes, please wait with reviewing until then.

**Comment by jessica-mitchell** [comment]
> @jessica-mitchell I would like to provide links between [rst documentation (bottom of file)](https://github.com/heplesser/nest-simulator/blob/253986423a9bbf4f3ede8d70dc8e5d751af7ee20/doc/htmldoc/neurons/exact-integration.rst) and [Jupyter documentation (top of notebook)](https://github.com/heplesser/nest-simulator/blob/253986423a9bbf4f3ede8d70dc8e5d751af7ee20/doc/htmldoc/neurons/model_details/IAF_Integration_Singularity.ipynb). Do you have any suggestion about how to best do that?

To link notebooks from restructured text files, you need to use the form 
```  
`IAF integration singularity <model_details/IAF_Integration_Singularity.ipynb>`_

```
Note that you can't just put the path in backticks,  it seems to require the angle brackets.

**Comment by heplesser** [comment]
@jougs @clinssen @jessica-mitchell I have now re-worked the class again, with a new approach to detecting cases in which the singularity causes problems. Please see the notebook before you look at the code (https://github.com/heplesser/nest-simulator/blob/354aa3ed3800c437a661513a36a30629e9dbdab8/doc/htmldoc/neurons/model_details/IAF_Integration_Singularity.ipynb). I also added a test and confirmed that all test cases fail if I deactivate the singularity avoidance mechanism. Un my laptop, `brunel_ps.sli` runs equally fast as on this branch as on master. Detailed results for ps-neurons change because the singluar propagators are now used under slightly different circumstances than before.

I am looking forward to your reviews.

**Comment by jougs** [comment]
@heplesser: can you please resolve conflicts?

**Comment by heplesser** [comment]
> @heplesser: can you please resolve conflicts?

Done, merging #2482 had mixed up some includes.

File: doc/htmldoc/neurons/model_details/IAF_Integration_Singularity.ipynb
Comment (jessica-mitchell) [inline doc/htmldoc/neurons/model_details/IAF_Integration_Singularity.ipynb:9]:
```suggestion
    "This notebook details how NEST handles the numerical instability of the exact integration propagator matrix $P = e^{A h}$ which arise if $\\tau_m\\approx \\tau_{\\text{syn}}$. For an overview over exact integration of integrate-and-fire neuron subthreshold dynamics, please see [exact-integration](model_details/exact-integration.rst).\n",
```

Comment (jessica-mitchell) [reply doc/htmldoc/neurons/model_details/IAF_Integration_Singularity.ipynb:9]: It's possible to link to rst files from notebooks  using the markdown syntax

Comment (heplesser) [reply doc/htmldoc/neurons/model_details/IAF_Integration_Singularity.ipynb]: Thanks for the hint, I have accepted it! Will this link to the rst file be re-written into a link to the corresponding HTML when documentation is rendered?

File: doc/htmldoc/neurons/exact-integration.rst
Comment (jougs) [inline doc/htmldoc/neurons/exact-integration.rst:188]:
```suggestion
The update is performed `here <https://github.com/nest/nest-simulator/blob/b3fc263e073f46f0732c10efb34fcc90f3b6771c/models/iaf_psc_alpha.cpp#L305-L307>`_. The first multiplication evolves the external input. The others are the multiplication of the matrix :math:`e^{Ah}` with :math:`y` (for inhibitory and excitatory inputs).
```

**Review by jougs — APPROVED** [review]
I mostly had a look at the code structure and this looks good to me. Regarding the propagator implementations itself, I trust the review of @clinssen.

File: libnestutil/iaf_propagator.h
Comment (heplesser) [reply libnestutil/iaf_propagator.h:42]: @jessica-mitchell Ok, I will check how to provide links for doxygen. And should I change here to the new directory layout or did you mean that you will fix it in #2453?

Comment (heplesser) [reply libnestutil/iaf_propagator.h:42]: @jessica-mitchell I'd suggest to keep the link to the new destination so we don't need to fix it very soon again. But how to insert a link properly into Doxygen documentation? I tried
```
<../doc/userdoc/model_details/IAF_neurons_singularity.ipynb>
```
based on https://www.doxygen.nl/manual/markdown.html#md_autolink, but that did not work (no error message either). Do you have a suggestion?

Comment (heplesser) [reply libnestutil/iaf_propagator.h:42]: Since the link above currently cannot work, I actually tried with <../doc/htmldoc/neurons/model_details/IAF_neurons_singularity.ipynb> with no luck.

Comment (jessica-mitchell) [reply libnestutil/iaf_propagator.h:42]: I also tried to create a link in doxygen, and the only thing I could get to work is a link to the URL [The link text](https://nest-simulator.readthedocs.io/en/latest/neurons/model_details/IAF_neurons_singularity.ipynb), which is not ideal.

Comment (heplesser) [reply libnestutil/iaf_propagator.h]: @jessica-mitchell I now changed the reference to
```
* @see `<NEST_SOURCE>/doc/userdoc/model_details/IAF_Integration_Singularity.ipynb`
```
This won't create a link, but should tell a developer enough to find the notebook. I think this is the best we can do at present.

**Comment by heplesser** [comment]
@clinssen Ping!

**Review by jessica-mitchell — APPROVED** [review]

File: doc/htmldoc/neurons/model_details/IAF_Integration_Singularity.ipynb
Comment (clinssen) [inline doc/htmldoc/neurons/model_details/IAF_Integration_Singularity.ipynb:726]: Not sure I follow plausibility argument; as far as I see it's actually the use of ``expm1()`` that saves us here (using ``exp() - 1`` exhibits similar issues as $P_{31}$).

**Review by clinssen — APPROVED** [review]
Approving with only a few minor comments. Cheers!

**Comment by heplesser** [comment]
@clinssen I am confused by the suggestions containing images. Could you clarify what you mean? In cells creating plots I tend to be generous with semicolons to suppress text output. Strictly speaking a semicolon is only needed after the final plot command in a cell, but tracking that during development is a bit tedious. I will review your comment on the plausibility argument.

File: doc/htmldoc/neurons/model_details/IAF_Integration_Singularity.ipynb
Comment (heplesser) [reply doc/htmldoc/neurons/model_details/IAF_Integration_Singularity.ipynb:726]: Yes, that is exactly my point: by using `expm1()` (which we also do in the C++ implementation in NEST), we end up with an expression for `P_32` which contains no differences at the C++ level and thus has no risk of introducing significant numerical noise by subtracting two almost identical numbers from each other. For `P_31` on the other hand, we have and explicit difference that we cannot avoid.

**Comment by heplesser** [comment]
@clinssen I merged the newest changes from master in the hope that all tests will pass again and beautified the notebook a bit more.

**Pull request merged**
